woocommerce/woocommerce-google-analytics-integration
# Implement consent mode v0

### Changes proposed in this Pull Request:


Add the minimalistic implementation to add support for [Google Consent mode](https://developers.google.com/tag-platform/security/guides/consent?consentmode=advanced)

We want to add MVP ASAP before Google starts penalizing merchants without this setup. (pcTzPl-25o-p2)
Later, we could iterate by adding settings to customize the snippet a bit using UI.

This PR adds a basic default and documents how to customize/extend it.

### Checks:


### Screenshots:
![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/00061cfd-deef-44f0-80e4-a346f92d2d7e)



### Detailed test instructions:


1. Buiild, Install and activate the extension
2. Deactivate GLA (if applicable, as the integration is broken for `remove-universal-analytics` branch)
3. Disable "Add to cart" events in settings (`/wp-admin/admin.php?page=wc-settings&tab=integration&section=google_analytics`) so WooCommerce Core will not set up its own `gtag` config making us do nothing.
4. Open [Tag Assistant](https://tagassistant.google.com/) 
5. Open shop page the page 
6. In Tag assistant you should see the "Consent" as the first entry.
7. If you're connecting from EEA region, you should see "Consent" being Denied by default
   ![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/361077b9-7e43-4edb-9107-2fad4ca9deb8)
8. If you're connecting from non-EEA region, you should see "Consent" not set 
   ![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/ad61fcf8-4377-4218-a4af-721047c24af2)
10. Check that Readme description
11. Add the snippet to customize the consent mode:
```php
add_action( 'wp_enqueue_scripts', function () {
    $code = "
        gtag( 'consent', 'default', {
            analytics_storage: 'granted',
            ad_storage: 'granted',
            ad_user_data: 'denied',
            ad_personalization: 'granted',
            region: 'PL', // <--------- SET IT TO YOUR REGION
        } );";

    wp_register_script( 'my-custom-consent-mode', '', array('woocommerce-google-analytics-integration'), null, [ 'in_footer' => true ] );
    wp_add_inline_script( 'my-custom-consent-mode', $code );
    wp_enqueue_script( 'my-custom-consent-mode' );
} );
```
Check that the second Consent entry is added with custom defaults
![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/08f72b9d-5a53-4c26-892e-d2b62bfe4a86)






### Additional details:
1. This customization capabilities are not ideal, as the is a change to get `gtag` hit before the second consent config is saved, as we may trigger page_view or other events. We may improve that in the future. 

   Especially since the current logic to force `main.js` to run after the document is ready is pretty hacky. (I guess that's why we add inline script` on `wp_footer` hook). I think we can make it in a cleaner fashion:
   - wrap all JS code that requires document ready in `DOMContentLoaded` event listener
   - execute the code that needs to run synchronously


### Changelog entry

> Add - Basic consent mode support.
